File: cheaptrips.java
import java.text.DecimalFormat;
import java.util.*;

public class Main {
	static List<Double> duration = new LinkedList<>();
	static List<Double> cost = new LinkedList<>();
	static HashMap<Integer, Double> partialCosts = new HashMap<>();
	static double[] discounts = { 1.00, 2.00, 4.00, 4.00, 4.00, 4.00 };
	static DecimalFormat df = new DecimalFormat("0.00");

	public static void main(String args[]) {
		Scanner scan = new Scanner(System.in);
		int n = scan.nextInt();
		double total = 0;
		int id = 0;
		boolean ok = true;
		for (int i = 0; i < n; i++) {
			double dur = scan.nextDouble();
			double cos = scan.nextDouble();
			duration.add(dur);
			cost.add(cos);
			if (dur > 120 && ok) {
				total += cos;
				id = i;
			} else {
				ok = false;
			}
		}
		if (id == (n - 1))
			System.out.println(String.valueOf(df.format(total)).replace(",", "."));
		else {
			double minCost = total + getMinCost(id);
			System.out.println(String.valueOf(df.format(minCost)).replace(",", "."));
		}

	}

	public static double getMinCost(int index) {
		if (index > cost.size())
			return 0;
		if (partialCosts.get(index) != null) {
			return partialCosts.get(index);
		}
		double partialDuration = 0;
		double partialCost = 0;
		double min = Double.MAX_VALUE;
		for (int i = 0; i < 6 && partialDuration < 120 && index + i < duration.size(); i++) {
			partialCost += cost.get(index + i) / discounts[i];
			if (index + i + 1 < duration.size()) {
				min = Math.min(min, partialCost + getMinCost(index + i + 1));
			} else
				min = partialCost;
			partialDuration += duration.get(index + i);
		}
		partialCosts.put(index, min);
		return min;
	}
}

File: cheaptrips.cpp
#include <bits/stdc++.h>

using namespace std;

long long dp[10001];
long long discount[] = {1, 2, 4, 4, 4, 4};
vector<long long> duration(10001), cost(10001);

long long getMinCost(int index) {
  if (index >= int(duration.size())) return 0;

  if (dp[index] != -1) return dp[index];

  long long minCost = LLONG_MAX;
  long long partialCost = 0;
  long long partialDuration = 0;

  for (int i = 0; partialDuration < 120 && i < 6 &&
                  (index + i) < int(duration.size())
                ; i++) {
    partialCost += cost[index + i]*100/discount[i];
    minCost = min(minCost, partialCost + getMinCost(index + i + 1));

    partialDuration += duration[index + i];
  }
  dp[index] = minCost;
  return minCost;
}

int main() {

  int n;
  cin >> n;
  for (int i = 0; i < n; i++) {
    cin >> duration[i] >> cost[i];
  }

  memset(dp, -1, sizeof dp);
  long long minCost = getMinCost(0);
  double resp = (double) minCost/100;
  cout << std::fixed;
  cout << std::setprecision(2);
  cout << resp << endl;
  return 0;
}
